retrier/retrier_test.go

```diff
@@ -153,6 +153,44 @@ func TestRetrierRunFnWithInfinite(t *testing.T) {
 	}
 }
 
+func TestRetrierCtxSurfaceWorkErrors(t *testing.T) {
+	// Will timeout before getting to errBar.
+	ctx, cancel := context.WithTimeout(context.Background(), 7*time.Millisecond)
+	defer cancel()
+	r := New([]time.Duration{0, 10 * time.Millisecond}, nil).WithSurfaceWorkErrors()
+	errExpected := []error{errFoo, errFoo, errBar, errBaz}
+	retries := 0
+	err := r.RunCtx(ctx, func(ctx context.Context) error {
+		if retries >= len(errExpected) {
+			return nil
+		}
+		err := errExpected[retries]
+		retries++
+		return err
+	})
+	if err != errFoo {
+		t.Error(err)
+	}
+}
+
+func TestRetrierRunFnWithSurfaceWorkErrors(t *testing.T) {
+	// Will timeout before getting to errBar.
+	ctx, cancel := context.WithTimeout(context.Background(), 7*time.Millisecond)
```

Comment:
Rather than using a timeout here (which is probably reliable enough, but best to avoid time-dependent testing if possible), I think we can manually cancel the context at the appropriate moment, for a more deterministic test.